doc/services/rtio/index.rst

```diff
@@ -0,0 +1,417 @@
+submission queue entry. However, cancellation as an API induces many potential
+complexities that might not be appropriate. It's something to be decided upon.
+
+Userspace Support
+=================
+
+RTIO with userspace is certainly plausible but would require the equivalent of
+a memory map call to map the shared ringbuffers and also potentially dma buffers.
+
+Additionally a DMA buffer interface would likely need to be provided for
+coherence and MMU usage.
+
+IODev and Executor API
+======================
+
+Lastly the API between an executor and iodev is incomplete.
+
+There are certain interactions that should be supported. Perhaps things like
+expanding a submission queue event into multiple submission queue entries in
+order to split up work that can be done by a device and work that can be done
+by a DMA controller.
+
+In some SoCs only specific DMA channels may be used with specific devices. In
+others there are requirements around needing a DMA handshake or specific
+triggering setups to tell the DMA when to start its operation.
+
+None of that, from the outward facing API, is an issue.
+
+It is however an unresolved task and issue from an internal API between the
+executor and iodev. This requires some SoC specifics and enabling those
+generically isn't likely possible. That's ok, an iodev and dma executor should
+be vendor specific, but an API needs to be there between them that is not!
+
+
+Special Hardware: Intel HDA
+===========================
+
+In some cases there's a need to always do things in a specific order
+with a specific buffer allocation strategy. Consider a DMA that *requires*
+the usage of a circular buffer segmented into blocks that may only be
+transferred one after another. This is the case of the Intel HDA stream for
+audio.
+
+In this scenario the above API can still work, but would require an additional
+buffer allocator to work with fixed sized segments.
+
+
+
+When to Use
+***********
+
+It's important to understand when DMA like transfers are useful and when they
+are not. It's a poor idea to assume that something made for high throughput will
+work for you. There is a computational, memory, and latency cost to setup the
+description of transfers.
+
+Polling at 1Hz an air sensor will almost certainly result in a net negative
+result compared to ad-hoc sensor (i2c/spi) requests to get the sample.
+
+Continuous transfers, driven by timer or interrupt, of data from a peripherals
+on board FIFO over I2C, I3C, SPI, MIPI, I2S, etc... maybe, but not always!
+
+Examples
+********
+
+Examples speak loudly about the intended uses and goals of an API. So several key
+examples are presented below. Some are entirely plausible today without any a
+big leap. Others (the sensor example) would require additional work in other
+APIs outside of RTIO as a sub system and are theoretical.
+
+Chained Blocking Requests
+=========================
+
+.. code-block:: C
+
```

Comment:
Adding comments through out the examples to make reading easier on the reader would be advised.  Assume your reader knows how to code but has no idea what any of these functions are doing. Adding a general concept on each block would make all the difference in the world.

Reply:
I agree, I will go back through these samples and update them to the current API with commentary